### PR TITLE
Add type hints

### PR DESCRIPTION
**Pull request by HebaruSan**
This is the same as #269 except it uses a branch in the KSP-SpaceDock/SpaceDock repo so the GitHub action will run (we hope!).

## Motivation

Type hints allow us to specify how a function is supposed to be used and catch when it isn't. They're especially helpful when writing new code, as trivial errors such as trying to access `my_dict['prop']` as `my_dict.prop` (my personal bugbear) are caught before run-time, but are also useful if you just need to know what a function returns or what to pass to it. They're a standard part of Python:

- https://www.python.org/dev/peps/pep-0484/

The `mypy` utility uses type hints to perform type checking:

- http://mypy-lang.org/

## Changes

- A `mypy.ini` file is added to the project root with some convenient defaults
- `.mypy_cache` is now git-ignored
- Type hints are added everywhere
- A GitHub Action is created to run mypy for commits and pull requests

```
$ mypy KerbalStuff 
Success: no issues found in 25 source files
```

### Possibly functional changes

The process of adding type hints revealed a few problems and opportunities for improvement.

- Unpublished mods can now be downloaded by admins, as per discussion in Discord
- Unreachable code fixed; CKAN triggers weren't being sent after mod authors marked a mod compatible with the latest game version:
  https://github.com/KSP-SpaceDock/SpaceDock/blob/e91894465c3a2885531355b9c813493ad93eaca4/KerbalStuff/blueprints/mods.py#L506-L507
- Removed handling of non-existent `templates/mobile` directory:
  https://github.com/KSP-SpaceDock/SpaceDock/tree/alpha/templates
  https://github.com/KSP-SpaceDock/SpaceDock/blob/b7cb750b5b71e497f5d791c1754e55c1ac90ed0c/KerbalStuff/app.py#L203-L207
- Cleaned up GitHUB OAuth helper function that could return redirects and strings, but was used as if it would never do that:
  https://github.com/KSP-SpaceDock/SpaceDock/blob/b7cb750b5b71e497f5d791c1754e55c1ac90ed0c/KerbalStuff/blueprints/login_oauth.py#L61-L72
  https://github.com/KSP-SpaceDock/SpaceDock/blob/b7cb750b5b71e497f5d791c1754e55c1ac90ed0c/KerbalStuff/blueprints/login_oauth.py#L134
  https://github.com/KSP-SpaceDock/SpaceDock/blob/b7cb750b5b71e497f5d791c1754e55c1ac90ed0c/KerbalStuff/blueprints/login_oauth.py#L161

Otherwise everything should work the same as before.

### CODE REVIEW

**Comment by HebaruSan** [comment]
@DasSkelett @RockyTV The GitHub action ran that time!

image

**Comment by HebaruSan** [comment]
I'm concerned about the mobile thing. I can find no trace of such files, but I have seen what might be screenshots of it. Does anyone know what's up with mobile? Is there some kind of automated tool generating those templates? A separate repo somewhere? @V1TA5?

**Comment by HebaruSan** [comment]
And now for my grand defense of "why `mypy`":

- It works
- It's easy
- It's the main thing that comes up if you search for Python type hint checkers
- It's hosted at https://github.com/python/mypy, the "python" part of which makes it look pretty official
- It's easy to switch to something else if we have a reason to

**Comment by DasSkelett** [comment]
> Is there some kind of automated tool generating those templates?

I don't think so. I searched my local repository and the running docker containers, and I can't find any clue of this folder.
I suspect it's a leftover of a previous attempt to create a mobile-friendly UI.

> but I have seen what might be screenshots of it.

Do you (still) have these screenshots?

**Comment by HebaruSan** [comment]
> Do you (still) have these screenshots?

See #130.

**Comment by DasSkelett** [comment]
> See #130.

Interesting. But I think this is no more. I can't find anything that would create this folder or something.

File: KerbalStuff/blueprints/admin.py [hunk 1]
```diff
@@ -1,8 +1,10 @@
 import math
+from typing import Union, List, Tuple
 
 from flask import Blueprint, render_template, redirect, request, abort, url_for
 from flask_login import login_user, current_user
-from sqlalchemy import desc, or_, func
+from sqlalchemy import desc, or_, func, Query
```
Comment (DasSkelett) [inline KerbalStuff/blueprints/admin.py:6]:
```suggestion
from sqlalchemy import desc, or_, func
from sqlalchemy.orm import Query
```

File: KerbalStuff/common.py [hunk 2]
```diff
@@ -3,20 +3,23 @@
 import urllib.parse
 import os
 from functools import wraps
+from typing import Union, List, Dict, Any, Optional, Callable, Tuple, Iterable
 
 from flask import jsonify, redirect, request, Response, abort, session
 from flask_login import current_user
 from werkzeug.utils import secure_filename
+import werkzeug.wrappers
+from sqlalchemy import Query
```
Comment (DasSkelett) [inline KerbalStuff/common.py:12]:
```suggestion
from sqlalchemy.orm import Query
```

File: KerbalStuff/blueprints/login_oauth.py [hunk 3]
```diff
@@ -1,11 +1,14 @@
...
 import os
 from collections import OrderedDict
+from typing import List, Dict, Optional, Union, Tuple, Any
+import werkzeug.wrappers
 
 from flask import Blueprint, render_template, request, redirect, session, jsonify, url_for, \
     current_app
 from flask_login import current_user, login_user
 from flask_oauthlib.client import OAuth
+from flask_oauth import OAuthRemoteApp
```
Comment (DasSkelett) [inline KerbalStuff/blueprints/login_oauth.py:11]:
```suggestion
from flask_oauthlib.client import OAuth, OAuthRemoteApp
```

**Comment by HebaruSan** [comment]
Rebased, resolved conflicts, fixed comments, added type hints for new code.

File: KerbalStuff/blueprints/admin.py (hunk 1 above)
Comment (HebaruSan) [reply KerbalStuff/blueprints/admin.py]: Thanks, did these as part of rebasing.

**Comment by HebaruSan** [comment]
I'm confused about the `OrderedDict` point. Is there an exception being thrown there?

**Comment by DasSkelett** [comment]
> I'm confused about the `OrderedDict` point. Is there an exception being thrown there?

Whoops, sorry, accidentally cut that off the stack trace when I pasted it. Here it is, still exists with the latest change :/
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/opt/spacedock/KerbalStuff/app.py", line 25, in <module>
    from .blueprints.login_oauth import list_defined_oauths, login_oauth
  File "/opt/spacedock/KerbalStuff/blueprints/login_oauth.py", line 22, in <module>
    DEFINED_OAUTHS: Optional[OrderedDict[str, Dict[str, Any]]] = None
TypeError: 'type' object is not subscriptable
```

**Comment by HebaruSan** [comment]
> Here it is, still exists with the latest change

Huh, that's weird. When are you seeing that, when running `mypy`? I don't think the type hints would have an effect otherwise, and I'm having no issues:

```
$ mypy KerbalStuff/
Success: no issues found in 25 source files
```

I did see some discussion of `OrderedDict` being added somewhat recently. How new is your `python`?

**Comment by HebaruSan** [comment]
Rebased on alpha and hopefully fixed the OrderedDict thing.

**Comment by HebaruSan** [comment]
Rebased and resolved conflicts again. The tests from #280 caught a bug in `/api/browse`.

**Pull request merged**